### PR TITLE
feat(move): #DRIV-45 add move nextcloud option

### PR DESCRIPTION
**Pull request by aymericdevriesere**
## Describe your changes
Context: We wanted to add the possibility to move a file within nextcloud's forlders.

https://user-images.githubusercontent.com/101819152/176392864-caa246ab-a367-46c1-a8ea-92664bbc4911.mp4

It is now possible to move a nextcloud file into a nextcloud folder. It is working as the rename option.
If the file name already exists in the target folder, the operation will fail.
## Checklist tests

- [x] Move one file into a folder
- [x] Move multiples files into a folder

## Issue ticket number and link
[ DRIV-45 ]
https://entsupport.gdapublic.fr/projects/DRIV/issues/DRIV-45
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java [hunk 1]
```diff
@@ -185,15 +186,15 @@ public Future<HttpResponse<Buffer>> getFiles(UserNextcloud.TokenProvider userSes
     }
 
     @Override
-    public Future<JsonObject> moveDocument(UserNextcloud.TokenProvider userSession, String path, String destPath) {
+    public Future<JsonObject> moveDocument(UserNextcloud.TokenProvider userSession, String path, String destPath, NextcloudMoveAction moveAction) {
         Promise<JsonObject> promise = Promise.promise();
         String endpoint = nextcloudConfig.host() + nextcloudConfig.webdavEndpoint() + "/" + userSession.userId();
-        this.listFiles(userSession, destPath)
+        resolveFilesForMove(userSession, destPath, moveAction)
```
Comment (aymericdevriesere) [inline src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java:192]: On aurait pu ajouter l'option de renommer le fichier si déjà existant

File: src/main/resources/public/ts/services/nextcloud.service.ts [hunk 2]
```diff
@@ -39,7 +39,7 @@ export const nextcloudService: INextcloudService = {
     },
 
     moveDocument: (userid: string, path: string, destPath: string): Promise<AxiosResponse> => {
-        const urlParam: string = `?path=${path}&destPath=${destPath}`
+        let urlParam: string = `?path=${path}&destPath=${destPath}`;
```
Comment (BastienBenard) [inline src/main/resources/public/ts/services/nextcloud.service.ts:42]: Pourquoi le passer la variable en let ?

File: src/main/java/fr/openent/nextcloud/core/constants/Field.java [hunk 3]
```diff
@@ -51,6 +51,7 @@ public class Field {
     public static final String URL = "url";
 
 
+
```
Comment (flmartineau) [inline src/main/java/fr/openent/nextcloud/core/constants/Field.java:54]: Ca me semble dispensable ça ^^'

File: src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts [hunk 4]
```diff
@@ -155,38 +155,79 @@ class ViewModel implements IViewModel {
...
+        if (folderContent.content instanceof SyncDocument && folderContent.content.isFolder) {
+            // if interacted into nextcloud
+            this.processMoveToNextcloud(document, folderContent.content, selectedFolderFromNextcloudTree);
+        }
+    }
+
+    private processMoveToWorkspace(folderContent: any, document: SyncDocument, selectedFolderFromNextcloudTree: SyncDocument): void {
+        if (folderContent.folder instanceof models.Element) {
+            let fileToMove: Set<SyncDocument> = new Set(this.selectedDocuments).add(document);
+            let finalUpload: Array<string> = Array.from(fileToMove)
```
Comment (BastienBenard) [inline src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts:172]: const filesPathUpload me semble plus parlant non ?

File: src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts [hunk 5]
```diff
@@ -155,38 +155,79 @@ class ViewModel implements IViewModel {
...
-                            console.error(message + err.message);
-                        });
-                }
-            }
-            if (folderContent.folder instanceof SyncDocument) {
-                // if interacted into nextcloud
+            this.processMoveToWorkspace(folderContent, document, selectedFolderFromNextcloudTree);
+        }
+        // if interacted into nextcloud (folderContent as being te targeted path to move in)
+        if (folderContent.content instanceof SyncDocument && folderContent.content.isFolder) {
```
Comment (flmartineau) [inline src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts:163]: Et si `folderContent `est null ? Ajoute la vérif ici aussi

File: src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts [hunk 6]
```diff
@@ -155,38 +155,79 @@ class ViewModel implements IViewModel {
...
+    }
+
+    private processMoveToWorkspace(folderContent: any, document: SyncDocument, selectedFolderFromNextcloudTree: SyncDocument): void {
+        if (folderContent.folder instanceof models.Element) {
+            let fileToMove: Set<SyncDocument> = new Set(this.selectedDocuments).add(document);
+            let finalUpload: Array<string> = Array.from(fileToMove)
+                .map((file: SyncDocument) => file.path);
+            if (finalUpload.length) {
+                this.nextcloudService.moveDocumentNextcloudToWorkspace(model.me.userId, finalUpload, folderContent.folder._id)
+                    .then(async (_: AxiosResponse) => {
```
Comment (BastienBenard) [inline src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts:176]: Je ne comprend pas le async (_: AxiosResponse). Est ce utile de mettre en async ?

File: src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts [hunk 7]
```diff
@@ -155,38 +155,79 @@ class ViewModel implements IViewModel {
...
+                        const message: string = "Error while attempting to move nextcloud document to workspace " +
+                            "or update nextcloud list";
+                        console.error(message + err.message);
+                    });
             }
         }
     }
 
+    private async moveAllDocuments(document: SyncDocument, target: SyncDocument): Promise<AxiosResponse[]> {
+        let promises: Array<Promise<AxiosResponse>> = [];
```
Comment (BastienBenard) [inline src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts:197]: const promises

File: src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts [hunk 8]
```diff
@@ -155,38 +155,79 @@ class ViewModel implements IViewModel {
...
+        this.selectedDocuments.push(document);
+        let selectedSet: Set<SyncDocument> = new Set(this.selectedDocuments.filter((doc: SyncDocument) => !doc.isFolder));
```
Comment (BastienBenard) [inline src/main/resources/public/ts/sniplets/content/workspace-nextcloud-content.sniplet.ts:199]: const selectedSet

**Review by BastienBenard — COMMENTED** [review]
T'es tests ne passe pas ;)

**Pull request merged**